File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.py
class Solution:
    def reverseWords(self, s: str) -> str:
        s=s.strip()
        ans=[]
        temp=[]
        i=len(s)-1
        while i>=0:
            while(i>=0 and s[i]!=' '):
                temp.append(s[i])
                i-=1
            # now we encounter with space
            while(i>=0 and s[i]==' '):
                i-=1
            if(len(ans)==0):
                ans.append(''.join(reversed(temp)))
            else:
                ans.append(' ')
                ans.append(''.join(reversed(temp)))
            temp=[]
        return ''.join(ans)

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        s=s.trim();
        StringBuilder ans = new StringBuilder("");
        StringBuilder sbt = new StringBuilder("");
        int i=s.length()-1;
        while(i>=0){
            while(i>=0 && s.charAt(i)!=' '){
                sbt.append(s.charAt(i));
                i--;
            }
            // now next index value is space
            while(i>=0 && s.charAt(i)==' ') i--;
            if(ans.length()==0){
                // ans is emply
                ans.append(sbt.reverse());
            }else{
                // before adding any word to ans we first add the space
                ans.append(' ');
                ans.append(sbt.reverse());
            }
            sbt = new StringBuilder("");
        }
        return ans.toString();

    }
}